test/15_target-hostapp.ts

```diff
@@ -433,47 +449,85 @@ export default () => {
 				expect(body.d[0]['os_variant']).to.be.not.null;
 			});
 
-			it('should provision with an invalidated hostapp release', async () => {
-				const devicePatchBody = {
-					local: {
-						os_version: 'balenaOS 2.52.0+rev1',
-						os_variant: 'prod',
-					},
-				};
-
-				await invalidatedReleaseDevice.patchStateV2(devicePatchBody);
-				const { body } = await supertest(admin)
-					.get(
-						`/${version}/device(${invalidatedReleaseDevice.id})?$select=should_be_operated_by__release`,
-					)
-					.expect(200);
-				expect(body.d[0]).to.not.be.undefined;
-				expect(body.d[0]['should_be_operated_by__release'].__id).to.equal(
-					invalidatedReleaseId,
-				);
-				expect(body.d[0]['os_version']).to.be.not.null;
-				expect(body.d[0]['os_variant']).to.be.not.null;
-				const supervisorVersion = 'v12.3.5';
-				const devicePatchBody2 = {
-					local: {
-						supervisor_version: supervisorVersion,
-					},
-				};
+			(
+				[
+					[
+						'release_tag only',
+						'balenaOS 2.52.0+rev1',
+						() => invalidatedTagOnlyReleaseId,
+					],
+					[
+						'semver only',
+						'balenaOS 2.52.1+rev1',
+						() => invalidatedSemverOnlyReleaseId,
+					],
+				] as const
+			).forEach(([osTypeTitlePart, initialOsVersion, getHostAppReleaseId]) => {
+				it(`should provision with an invalidated ${osTypeTitlePart} hostapp release`, async () => {
+					const invalidatedReleaseDevice = await fakeDevice.provisionDevice(
+						admin,
+						applicationId,
+					);
+					const initialInvalidatedReleaseId = getHostAppReleaseId();
+					await invalidatedReleaseDevice.patchStateV2({
+						local: {
+							os_version: initialOsVersion,
+							os_variant: 'prod',
+						},
+					});
+					await expectResourceToMatch(
+						pineUser,
+						'device',
+						invalidatedReleaseDevice.id,
+						{
+							os_version: initialOsVersion,
+							os_variant: 'prod',
+							should_be_operated_by__release: {
+								__id: initialInvalidatedReleaseId,
+							},
+						},
+					);
 
-				await invalidatedReleaseDevice.patchStateV2(devicePatchBody2);
-				// after provisioning to our invalidated release, let's make sure we're not blocked
-				// in further PATCHing (using supervisor_version here as a proxy/dummy value)
-				const resp = await supertest(admin)
-					.get(
-						`/${version}/device(${invalidatedReleaseDevice.id})?$select=supervisor_version`,
-					)
-					.expect(200);
-				expect(resp.body.d[0]['supervisor_version']).to.equal(
-					supervisorVersion,
-				);
+					const supervisorVersion = 'v12.3.5';
+					const newOsVersion = 'balenaOS 2.88.5+rev1';
+					await invalidatedReleaseDevice.patchStateV2({
+						local: {
+							supervisor_version: supervisorVersion,
+							os_version: newOsVersion,
+							os_variant: 'production',
```

Comment:
Doubt it matters, but we don't actually have any "os_variant=production" device records in our production db. Maybe you meant "prod" here?